src/libcore/num/mod.rs

```diff
@@ -112,6 +112,112 @@ nonzero_integers! {
     #[stable(feature = "signed_nonzero", since = "1.34.0")] NonZeroIsize(isize);
 }
 
+/// An error which can be returned when parsing a non-zero integer.
+///
+/// # Potential causes
+///
+/// Among other causes, `ParseNonZeroIntError` can be thrown because of leading or trailing
+/// whitespace in the string e.g., when it is obtained from the standard input.
+/// Using the [`str.trim()`] method ensures that no whitespace remains before parsing.
+///
+/// [`str.trim()`]: ../../std/primitive.str.html#method.trim
+#[unstable(feature = "nonzero_parse", issue = "0")]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseNonZeroIntError {
+    kind: NonZeroIntErrorKind,
+}
+
+/// Enum to store the various types of errors that can cause parsing a non-zero integer to fail.
+#[unstable(feature = "nonzero_parse", issue = "0")]
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum NonZeroIntErrorKind {
+    /// Value being parsed is empty.
+    ///
+    /// Among other causes, this variant will be constructed when parsing an empty string.
+    Empty,
+    /// Contains an invalid digit.
+    ///
+    /// Among other causes, this variant will be constructed when parsing a string that
+    /// contains a letter.
+    InvalidDigit,
+    /// Integer is too large to store in target integer type.
+    Overflow,
+    /// Integer is too small to store in target integer type.
+    Underflow,
+    /// Integer contains the value `0` which is forbidden for a non-zero integer
+    Zero,
+}
+
+#[unstable(feature = "nonzero_parse", issue = "0")]
+impl From<ParseIntError> for ParseNonZeroIntError {
+    fn from(p: ParseIntError) -> Self {
+        use self::IntErrorKind as IK;
+        use self::NonZeroIntErrorKind as NK;
+        ParseNonZeroIntError {
+            kind: match p.kind {
+                IK::Empty => NK::Empty,
+                IK::InvalidDigit => NK::InvalidDigit,
+                IK::Overflow => NK::Overflow,
+                IK::Underflow => NK::Underflow,
+            },
+        }
+    }
+}
+
+impl ParseNonZeroIntError {
+    /// Outputs the detailed cause of parsing an integer failing.
+    #[unstable(feature = "int_error_matching",
+               reason = "it can be useful to match errors when making error messages \
+                         for integer parsing",
+               issue = "22639")]
+    pub fn kind(&self) -> &NonZeroIntErrorKind {
+        &self.kind
+    }
+
+    #[unstable(feature = "int_error_internals",
+               reason = "available through Error trait and this method should \
+                         not be exposed publicly",
+               issue = "0")]
+    #[doc(hidden)]
+    pub fn __description(&self) -> &str {
```

Comment:
There should be an accompanying Error implementation in libstd, right?

Reply:
This means something similar to this: (which is why the `__description` method is public.)

https://github.com/rust-lang/rust/blob/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libstd/error.rs#L474-L479

Reply:
Okay, I was wondering, because afaik the `ParseIntError` doesn't implement `Error`

Reply:
It… does? Exactly where I linked, in the comment you’re responding to.

Reply:
wow. I'm blind ^^ sorry